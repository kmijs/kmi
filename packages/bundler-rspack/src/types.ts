import type {
  PluginItem as BabelPlugin,
  TransformOptions as BabelTransformOptions,
} from '@kmijs/bundler-compiled/compiled/babel/core'
import type { KmiTarget } from '@kmijs/bundler-shared'
import { Env, type IBundlerConfig } from '@kmijs/bundler-shared-config'
import type {
  LightningcssLoaderOptions,
  SwcLoaderOptions,
} from '@kmijs/bundler-shared/rspack'

export enum Transpiler {
  babel = 'babel',
  swc = 'swc',
  none = 'none',
}

export { Env }

export interface DeadCodeParams {
  patterns?: string[]
  exclude?: string[]
  failOnHint?: boolean
  detectUnusedFiles?: boolean
  detectUnusedExport?: boolean
  context?: string
}

export type ExtraSwcPlugins = Array<[string, any]>

export type ModifySwcLoaderOptions = (
  options: SwcLoaderOptions,
  opts: {
    env: Env
    target: KmiTarget
    isServer: boolean
    isWebWorker: boolean
  },
) => SwcLoaderOptions

export interface IConfig extends IBundlerConfig {
  cssMinifierOptions?: { [key: string]: any }
  depTranspiler?: `${Transpiler}`
  srcTranspiler?: `${Transpiler}`
  srcTranspilerOptions?: Record<string, any>
  svgr?: { [key: string]: any }
  svgo?: { [key: string]: any } | false
  swc?: SwcLoaderOptions
  extraSwcPlugins?: ExtraSwcPlugins
  modifySwcLoaderOptions?: ModifySwcLoaderOptions
  lightningcssLoader?: boolean | LightningcssLoaderOptions
  [key: string]: any
}

export type BabelLoaderOptions = BabelTransformOptions & {
  /**
   * When set, the given directory will be used to cache the results of the loader.
   */
  cacheDirectory?: string | boolean
  /**
   * Can be set to a custom value to force cache busting if the identifier changes.
   */
  cacheIdentifier?: string
  /**
   * When set, each Babel transform output will be compressed with Gzip.
   */
  cacheCompression?: boolean
  /**
   * The path of a module that exports a custom callback.
   */
  customize?: string | null
  /**
   * Takes an array of context function names. E.g.
   */
  metadataSubscribers?: string[]
}

type OneOrMany<T> = T | T[]
export type ConfigChainWithContext<T, Ctx> = OneOrMany<
  T | ((config: T, ctx: Ctx) => T | void)
>

export interface PresetTypescriptOptions {
  ignoreExtensions?: boolean
  allowDeclareFields?: boolean
  allowNamespaces?: boolean
  disallowAmbiguousJSXLike?: boolean
  jsxPragma?: string
  jsxPragmaFrag?: string
  onlyRemoveTypeImports?: boolean
  optimizeConstEnums?: boolean
  rewriteImportExtensions?: boolean

  allExtensions?: boolean
  isTSX?: boolean
}

export type BabelConfigUtils = {
  addPlugins: (plugins: BabelPlugin[]) => void
  addPresets: (presets: BabelPlugin[]) => void
  removePlugins: (plugins: string | string[]) => void
  removePresets: (presets: string | string[]) => void
  modifyPresetTypescriptOptions: (options: PresetTypescriptOptions) => void
  modifyPluginOptions: (
    pluginName: string,
    options: Record<string, any>,
  ) => void
  isWebWorker?: boolean
  isServer?: boolean
}

export interface IExtraRspackOpts {
  extraSwcPlugins?: ExtraSwcPlugins
  modifySwcLoaderOptions?: ModifySwcLoaderOptions
}
