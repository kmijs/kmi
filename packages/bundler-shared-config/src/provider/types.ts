import type { BabelPlugin } from '@kmijs/bundler-compiled'
import type { KmiTarget } from '@kmijs/bundler-shared'
import type {
  Application,
  RequestHandler,
} from '@kmijs/bundler-shared/compiled/express'
import type {
  Compiler,
  Configuration,
  MultiCompiler,
} from '@kmijs/bundler-shared/rspack'
import type Config from '@kmijs/bundler-shared/rspack-chain'
import type { Env, IUserConfig } from '../types'

export interface IBaseConfigOpts<U extends IUserConfig = IUserConfig> {
  cwd: string
  rootDir?: string
  entry: Record<string, string>
  config: U
  babelPreset?: BabelPlugin[]
  beforeBabelPlugins?: BabelPlugin[]
  beforeBabelPresets?: BabelPlugin[]
  extraBabelPlugins?: BabelPlugin[]
  extraBabelPresets?: BabelPlugin[]
  disableCopy?: boolean
  cache?: any
  pkg?: any
  chainWebpack?: (config: any, opts: any) => void
  modifyWebpackConfig?: (config: any) => any
  /**
   * onBeforeCreateCompiler 是在创建底层 Compiler 实例前触发的回调函数
   * 你可以通过 bundlerConfigs 参数获取到 Rspack 配置数组，数组中可能包含一份或多份 Rspack(Webpack) 配置，这取决于是否配置了多个 environments。
   * @param args
   * @returns
   */
  onBeforeCreateCompiler?: (...args: any[]) => void
  /**
   * onAfterCreateCompiler 是在创建底层 Compiler 实例后触发的回调函数
   * 你可以通过 compiler 参数获取到 Compiler 实例对象
   * @param args
   * @returns
   */
  onAfterCreateCompiler?: (...args: any[]) => void
  /**
   * 构建目标
   * @default 'web'
   * web: 浏览器
   * node: 服务端
   * web-worker: 浏览器 worker
   */
  target?: KmiTarget
}

export interface IBaseServerOpts {
  onDevCompileDone?: (opts: {
    stats: any
    isFirstCompile: boolean
    time: number
  }) => void
  onProgress?: ({ progresses }: { progresses: any[] }) => void
  onAfterStartDevServer?: (opts: {
    protocol: string
    port: number
    host: string
  }) => Promise<void>
  onBeforeMiddleware?: (app: Application) => void
  beforeMiddlewares?: RequestHandler[]
  afterMiddlewares?: RequestHandler[]
  port?: number
  host?: string
  ip?: string
  hmr: boolean
}

export interface IDevOpts<U extends IUserConfig = IUserConfig>
  extends IBaseConfigOpts<U>,
    Omit<IBaseServerOpts, 'hmr'> {}

export interface IBuildOpts<U extends IUserConfig = IUserConfig>
  extends IBaseConfigOpts<U> {
  onBuildComplete?: (...args: any[]) => void
  clean?: boolean
  watch?: boolean
}

export type BundlerType = 'webpack' | 'rspack'

export interface IBaseBundlerConfigOpts<U extends IUserConfig = IUserConfig> {
  cwd: string
  rootDir?: string
  env: Env
  entry: Record<string, string>
  extraBabelPresets?: BabelPlugin[]
  extraBabelPlugins?: BabelPlugin[]
  extraBabelIncludes?: Array<string | RegExp>
  babelPreset?: BabelPlugin[]
  chainWebpack?: (config: Config, opts: any) => void
  modifyWebpackConfig?: (config: Configuration, opts: any) => Configuration
  hash?: boolean
  hmr?: boolean
  staticPathPrefix?: string
  userConfig: U
  analyze?: any
  name?: string
  cache?: {
    absNodeModulesPath?: string
    buildDependencies?: string[]
    cacheDirectory?: string
  }
  pkg?: Record<string, any>
  disableCopy?: boolean
  host?: string
  port?: number
  environment?: string
  /**
   * onBeforeCreateCompiler 是在创建底层 Compiler 实例前触发的回调函数
   * 你可以通过 bundlerConfigs 参数获取到 Rspack 配置数组，数组中可能包含一份或多份 Rspack(Webpack) 配置，这取决于是否配置了多个 environments。
   * @param args
   * @returns
   */
  onBeforeCreateCompiler?: (...args: any[]) => void
  /**
   * onAfterCreateCompiler 是在创建底层 Compiler 实例后触发的回调函数
   * 你可以通过 compiler 参数获取到 Compiler 实例对象
   * @param args
   * @returns
   */
  onAfterCreateCompiler?: (...args: any[]) => void
  /**
   * 构建目标
   * @default 'web'
   * web: 浏览器
   * node: 服务端
   * web-worker: 浏览器 worker
   */
  target: KmiTarget
}

export interface IBaseCreateServerOpts<U extends IUserConfig = IUserConfig>
  extends Omit<IBaseBundlerConfigOpts<U>, 'hmr'>,
    IBaseServerOpts {
  bundlerConfigs: Configuration | Configuration[]
  bundlerType: BundlerType
  createCompiler: (opts: any) => Compiler | MultiCompiler
}
