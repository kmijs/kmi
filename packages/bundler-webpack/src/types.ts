import type { IBundlerConfig } from '@kmijs/bundler-shared-config'
import type { TransformOptions as EsbuildOptions } from '@kmijs/bundler-shared/esbuild'

export { Env } from '@kmijs/bundler-shared-config'

export enum Transpiler {
  babel = 'babel',
  esbuild = 'esbuild',
  none = 'none',
}

export interface ICopy {
  from: string
  to: string
}

export interface DeadCodeParams {
  patterns?: string[]
  exclude?: string[]
  failOnHint?: boolean
  detectUnusedFiles?: boolean
  detectUnusedExport?: boolean
  context?: string
}

export interface IConfig extends IBundlerConfig {
  cssMinifierOptions?: { [key: string]: any }
  depTranspiler?: `${Transpiler}`
  srcTranspiler?: `${Transpiler}`
  srcTranspilerOptions?: ISrcTranspilerOpts
  svgr?: { [key: string]: any }
  svgo?: { [key: string]: any } | false
  babelLoaderCustomize?: string
  esbuildMinifyIIFE?: boolean
  checkDepCssModules?: boolean
  [key: string]: any
}

export interface ISrcTranspilerOpts {
  esbuild?: Partial<EsbuildOptions>
}

export interface ISwcPluginOpts {
  enableAutoCssModulesPlugin?: boolean
}
