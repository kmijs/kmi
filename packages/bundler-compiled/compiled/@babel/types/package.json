{"name":"@babel/types","version":"7.23.6","author":"The Babel Team (https://babel.dev/team)","license":"MIT","types":"./lib/index-legacy.d.ts"}
