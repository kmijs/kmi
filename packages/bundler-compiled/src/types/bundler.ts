import type { PluginItem as BabelPlugin } from '../../compiled/@babel/core'

export interface ICopy {
  from: string
  to: string
}

export type { BabelPlugin }

export enum CSSMinifier {
  esbuild = 'esbuild',
  cssnano = 'cssnano',
  lightningcss = 'lightningcss',
  none = 'none',
}

export enum JSMinifier {
  swc = 'swc',
  none = 'none',
  terser = 'terser',
  esbuild = 'esbuild',
}
