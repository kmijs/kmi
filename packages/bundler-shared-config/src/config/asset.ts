import { CHAIN_ID } from '@kmijs/bundler-shared'
import type { GeneratorOptionsByModuleType } from '@kmijs/bundler-shared/rspack'
import type Config from '@kmijs/bundler-shared/rspack-chain'
import { pathe } from '@kmijs/shared'
import {
  AUDIO_EXTENSIONS,
  FONT_EXTENSIONS,
  IMAGE_EXTENSIONS,
  VIDEO_EXTENSIONS,
} from '../constants'
import type { SharedConfigOptions } from '../types'
import { getFilename } from '../utils/getFilename'
import { getInlineLimit } from '../utils/getInlineLimit'
import { getOutputPath } from '../utils/getOutputPath'

const chainStaticAssetRule = ({
  emit,
  rule,
  maxSize,
  filename,
  assetType,
}: {
  emit: boolean
  rule: Config.Rule
  maxSize: number
  filename: string
  assetType: string
}) => {
  const generatorOptions:
    | GeneratorOptionsByModuleType['asset']
    | GeneratorOptionsByModuleType['asset/resource'] = {
    filename,
  }

  if (emit === false) {
    generatorOptions.emit = false
  }

  // force to url: "foo.png?url" or "foo.png?__inline=false"
  rule
    .oneOf(`${assetType}-asset-url`)
    .type('asset/resource')
    .resourceQuery(/(__inline=false|url)/)
    .set('generator', generatorOptions)

  // force to inline: "foo.png?inline"
  rule
    .oneOf(`${assetType}-asset-inline`)
    .type('asset/inline')
    .resourceQuery(/inline/)

  // default: when size < dataUrlCondition.maxSize will inline
  rule
    .oneOf(`${assetType}-asset`)
    .type('asset')
    .parser({
      dataUrlCondition: {
        maxSize,
      },
    })
    .set('generator', generatorOptions)
}

export function getRegExpForExts(exts: string[]): RegExp {
  const matcher = exts
    .map((ext) => ext.trim())
    .map((ext) => (ext.startsWith('.') ? ext.slice(1) : ext))
    .join('|')

  return new RegExp(
    exts.length === 1 ? `\\.${matcher}$` : `\\.(?:${matcher})$`,
    'i',
  )
}

export function applyAsset(opts: SharedConfigOptions) {
  const { config, userConfig } = opts

  const { emitAssets = true, assetsInclude } = userConfig

  const getMergedFilename = (
    assetType: 'svg' | 'font' | 'image' | 'media' | 'assets',
  ) => {
    const distDir = getOutputPath(opts, assetType)
    const filename = getFilename(opts, assetType)
    return pathe.join(distDir, filename)
  }

  const createAssetRule = (
    assetType: 'svg' | 'font' | 'image' | 'media',
    exts: string[],
    emit: boolean,
  ) => {
    const regExp = getRegExpForExts(exts)
    const maxSize = getInlineLimit(userConfig, assetType)
    const rule = config.module.rule(assetType).test(regExp)

    chainStaticAssetRule({
      emit,
      rule,
      maxSize,
      filename: getMergedFilename(assetType),
      assetType,
    })
  }

  // image
  createAssetRule(CHAIN_ID.RULE.IMAGE, IMAGE_EXTENSIONS, emitAssets)
  // svg
  createAssetRule(CHAIN_ID.RULE.SVG, ['svg'], emitAssets)
  // media
  createAssetRule(
    CHAIN_ID.RULE.MEDIA,
    [...VIDEO_EXTENSIONS, ...AUDIO_EXTENSIONS],
    emitAssets,
  )
  // font
  createAssetRule(CHAIN_ID.RULE.FONT, FONT_EXTENSIONS, emitAssets)
  // assets
  const assetsFilename = getMergedFilename('assets')
  config.output.assetModuleFilename(assetsFilename)
  if (!emitAssets) {
    config.module.generator.merge({ 'asset/resource': { emit: false } })
  }

  if (assetsInclude) {
    const rule = config.module
      .rule(CHAIN_ID.RULE.ADDITIONAL_ASSETS)
      .test(assetsInclude)

    const maxSize = getInlineLimit(userConfig, 'assets')

    chainStaticAssetRule({
      emit: emitAssets,
      rule,
      maxSize,
      filename: assetsFilename,
      assetType: 'additional-assets',
    })
  }
}
