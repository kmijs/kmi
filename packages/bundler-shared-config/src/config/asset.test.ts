import Config from '@kmijs/bundler-shared/rspack-chain'
import { describe, expect, test } from 'vitest'
import { Env, type SharedConfigOptions } from '../types'
import { applyAsset } from './asset'

const createConfig = (opts: Partial<SharedConfigOptions>) => {
  const config = new Config()
  applyAsset({
    cwd: __dirname,
    env: Env.development,
    config,
    ...opts,
  } as unknown as SharedConfigOptions)
  return config.toConfig()
}
describe('asset', () => {
  test('应该能够添加额外的资源文件', () => {
    const config = createConfig({
      userConfig: {
        assetsInclude: [/\.json5$/],
      },
      staticPathPrefix: 'static',
    })
    expect(config).toMatchSnapshot()
  })

  test('应该允许使用 outputPath.image 修改输出路径', async () => {
    const config = createConfig({
      userConfig: {
        outputPath: {
          image: 'foo',
        },
      },
      staticPathPrefix: 'static',
    })
    expect(config).toMatchSnapshot()
  })

  test('应该正确添加图片规则', async () => {
    const config = createConfig({
      userConfig: {
        outputPath: {
          image: '',
        },
      },
    })
    expect(config).toMatchSnapshot()
  })

  test('应该允许使用 filename.image 修改文件名', async () => {
    const config = createConfig({
      userConfig: {
        filename: {
          image: 'demo/img/foo[ext]',
        },
      },
    })
    expect(config).toMatchSnapshot()
  })

  test('应该能正确的配置 inlineLimit', async () => {
    const config = createConfig({
      userConfig: {
        inlineLimit: Number.MAX_SAFE_INTEGER,
      },
    })
    expect(config).toMatchSnapshot()
  })

  test('inlineLimit 允许单独设置', async () => {
    const config = createConfig({
      userConfig: {
        inlineLimit: {
          image: 5 * 1024,
          media: 1024,
          svg: 233,
        },
      },
    })
    expect(config).toMatchSnapshot()
  })
})
