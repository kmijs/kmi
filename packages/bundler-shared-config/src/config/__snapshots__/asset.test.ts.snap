// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`asset > inlineLimit 允许单独设置 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 5120,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:png\\|jpg\\|jpeg\\|pjpeg\\|pjp\\|gif\\|bmp\\|webp\\|ico\\|apng\\|avif\\|tif\\|tiff\\|jfif\\|cur\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 233,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 1024,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:mp4\\|webm\\|ogg\\|mov\\|mp3\\|wav\\|flac\\|aac\\|m4a\\|opus\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:woff\\|woff2\\|eot\\|ttf\\|otf\\|ttc\\)\\$/i,
      },
    ],
  },
  "output": {
    "assetModuleFilename": "[name][ext]",
  },
}
`;

exports[`asset > 应该允许使用 filename.image 修改文件名 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "generator": {
              "filename": "demo/img/foo[ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "demo/img/foo[ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:png\\|jpg\\|jpeg\\|pjpeg\\|pjp\\|gif\\|bmp\\|webp\\|ico\\|apng\\|avif\\|tif\\|tiff\\|jfif\\|cur\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:mp4\\|webm\\|ogg\\|mov\\|mp3\\|wav\\|flac\\|aac\\|m4a\\|opus\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:woff\\|woff2\\|eot\\|ttf\\|otf\\|ttc\\)\\$/i,
      },
    ],
  },
  "output": {
    "assetModuleFilename": "[name][ext]",
  },
}
`;

exports[`asset > 应该允许使用 outputPath.image 修改输出路径 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "generator": {
              "filename": "foo/[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "foo/[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:png\\|jpg\\|jpeg\\|pjpeg\\|pjp\\|gif\\|bmp\\|webp\\|ico\\|apng\\|avif\\|tif\\|tiff\\|jfif\\|cur\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/svg/[name].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/svg/[name].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/assets/[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/assets/[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:mp4\\|webm\\|ogg\\|mov\\|mp3\\|wav\\|flac\\|aac\\|m4a\\|opus\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/font/[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/font/[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:woff\\|woff2\\|eot\\|ttf\\|otf\\|ttc\\)\\$/i,
      },
    ],
  },
  "output": {
    "assetModuleFilename": "static/assets/[name][ext]",
  },
}
`;

exports[`asset > 应该正确添加图片规则 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:png\\|jpg\\|jpeg\\|pjpeg\\|pjp\\|gif\\|bmp\\|webp\\|ico\\|apng\\|avif\\|tif\\|tiff\\|jfif\\|cur\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/svg/[name].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/svg/[name].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/assets/[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/assets/[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:mp4\\|webm\\|ogg\\|mov\\|mp3\\|wav\\|flac\\|aac\\|m4a\\|opus\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/font/[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/font/[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:woff\\|woff2\\|eot\\|ttf\\|otf\\|ttc\\)\\$/i,
      },
    ],
  },
  "output": {
    "assetModuleFilename": "static/assets/[name][ext]",
  },
}
`;

exports[`asset > 应该能够添加额外的资源文件 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:png\\|jpg\\|jpeg\\|pjpeg\\|pjp\\|gif\\|bmp\\|webp\\|ico\\|apng\\|avif\\|tif\\|tiff\\|jfif\\|cur\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/[name].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/[name].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:mp4\\|webm\\|ogg\\|mov\\|mp3\\|wav\\|flac\\|aac\\|m4a\\|opus\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:woff\\|woff2\\|eot\\|ttf\\|otf\\|ttc\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "static/[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "static/[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 4096,
              },
            },
            "type": "asset",
          },
        ],
        "test": [
          /\\\\\\.json5\\$/,
        ],
      },
    ],
  },
  "output": {
    "assetModuleFilename": "static/[name][ext]",
  },
}
`;

exports[`asset > 应该能正确的配置 inlineLimit 1`] = `
{
  "module": {
    "rules": [
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 9007199254740991,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:png\\|jpg\\|jpeg\\|pjpeg\\|pjp\\|gif\\|bmp\\|webp\\|ico\\|apng\\|avif\\|tif\\|tiff\\|jfif\\|cur\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name].svg",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name].svg",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 9007199254740991,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.svg\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 9007199254740991,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:mp4\\|webm\\|ogg\\|mov\\|mp3\\|wav\\|flac\\|aac\\|m4a\\|opus\\)\\$/i,
      },
      {
        "oneOf": [
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "resourceQuery": /\\(__inline=false\\|url\\)/,
            "type": "asset/resource",
          },
          {
            "resourceQuery": /inline/,
            "type": "asset/inline",
          },
          {
            "generator": {
              "filename": "[name][ext]",
            },
            "parser": {
              "dataUrlCondition": {
                "maxSize": 9007199254740991,
              },
            },
            "type": "asset",
          },
        ],
        "test": /\\\\\\.\\(\\?:woff\\|woff2\\|eot\\|ttf\\|otf\\|ttc\\)\\$/i,
      },
    ],
  },
  "output": {
    "assetModuleFilename": "[name][ext]",
  },
}
`;
